### ISSUE TITLE
[Question] Authentication via auth_query

### ISSUE THREAD

**Issue by WagnerPMC** [postgresml/pgcat#606]
## Versions

PGCat: 1.1.1
PostgreSQL: Postgres Pro 15 (similar to PostgreSQL 15)

## Configurations
pgcat.toml
```
auth_query = "SELECT usename, passwd FROM public.pgcat_auth($1)"
auth_query_user = "pgcat_auth_user"
auth_query_password = ""
```

pg_hba.conf
```
hostssl all pgcat_auth_user 172.16.3.2/32 trust
```

## Description

My `pg_hba.conf` file allows the pgcat_auth_user user to be allowed into the server unhindered, but only to execute the `public.pgcat_auth` function. But something goes wrong when PGCat does it.

I'm catching the next error:
```
pgcat-pgcat-1  | 2023-09-26T23:05:53.212508Z  WARN ThreadId(04) pgcat: Client disconnected with error
 ClientAuthPassthroughError("AuthPassthroughError(\"Error trying to obtain password from auth_query, ignoring hash for user 'xxxxxxxxxx'. Error: ServerStartupError(\\\"error message\\\", ServerIdentifier { username: \\\"pgcat_auth_user\\\", database: \\\"xxxxxx\\\" })\")", ClientIdentifier { application_name: "pgcat", username: "xxxxxxx", pool_name: "xxxxxxxx" })
```

For some reason PGCat can't get a hash of my password.
To prove that I have everything set up correctly, I will show the following screenshot:
![issue-598](https://github.com/postgresml/pgcat/assets/97305733/c52dc720-6701-4b8b-b373-6bc46a06686b)

So what can it be?

**Comment by levkk** [comment]
Auth query only works with md5 which was removed in Postgres 14.

**Comment by WagnerPMC** [comment]
> Auth query only works with md5 which was removed in Postgres 14.

What can be done in such a case?

**Comment by levkk** [comment]
Only password auth through config is supported right now. If you want passthrough for Postgres 14+, we need to figure out if we can do passthrough with SCRAM or how to just store SCRAM strings in the config file instead of plain text passwords. Pgbouncer supports the latter, so we can take a look at their implementation for reference, but ideally, we figure out passthrough for SCRAM I think.

**Comment by WagnerPMC** [comment]
> Only password auth through config is supported right now. If you want passthrough for Postgres 14+, we need to figure out if we can do passthrough with SCRAM or how to just store SCRAM strings in the config file instead of plain text passwords. Pgbouncer supports the latter, so we can take a look at their implementation for reference, but ideally, we figure out passthrough for SCRAM I think.

Thanks, I got it.
Even though I'm new to Rust, I'll think about how it can be done and if I come up with something, I'll be sure to push it to PR.

**Comment by WagnerPMC** [comment]
Reading the PostgreSQL protocol documentation, I was still puzzled why your code crashes with an error, the reason for which is the inability to read the response buffer from the database. And it's not even about SCRAM.

I've modified the source code a bit to see what error occurs and the reason why PGCat can't read it from the buffer:
![shit](https://github.com/postgresml/pgcat/assets/97305733/7001741d-a792-4d31-bc29-caff7987f8bc)

In the original code, it happens here:
https://github.com/postgresml/pgcat/blob/a054b454d2566e1a90ef373c7a5360850f12484e/src/server.rs#L716-L726

This is happening because someone screwed up a little bit in that line:
https://github.com/postgresml/pgcat/blob/a054b454d2566e1a90ef373c7a5360850f12484e/src/server.rs#L716-L726

The solution is to subtract 5 bytes from the message length (len). (4 bytes are the message length + the length field itself, 1 byte is the message type).
Basically, you can see it in the screenshot.

I'll look into this fix myself, and see what I can do with SCRAM authentication

**Comment by WagnerPMC** [comment]
As I expected, I now get an error message from the DB server.
![shit2](https://github.com/postgresml/pgcat/assets/97305733/13c86dbe-f80d-4340-a189-9baafb1cab51)

Earlier, I was getting:
```
pgcat-pgcat-1  | 2023-09-26T23:05:53.212508Z  WARN ThreadId(04) pgcat: Client disconnected with error
 ClientAuthPassthroughError("AuthPassthroughError(\"Error trying to obtain password from auth_query, ignoring hash for user 'xxxxxxxxxx'. Error: ServerStartupError(\\\"error message\\\", ServerIdentifier { username: \\\"pgcat_auth_user\\\", database: \\\"xxxxxx\\\" })\")", ClientIdentifier { application_name: "pgcat", username: "xxxxxxx", pool_name: "xxxxxxxx" })
```

This was due to the end of the buffer being reached earlier than planned. I removed 5 bytes from the plan, now the buffer reads properly and I can get a normal error message.

**Comment by WagnerPMC** [comment]
@levkk , please give a comment

**Comment by WagnerPMC** [comment]
> Seems reasonable. A PR is welcome, if you have a moment to provide a fix.

I will prepare a PR, but in the meantime please consider my other PR.

**Comment by WagnerPMC** [comment]
So far I'm about halfway to end-to-end SCRAM-SHA-256 authentication. Looking at your code, I realize what nonsense it is. I really hope that you will not abandon your project, it really has a lot of room to grow!

**Comment by levkk** [comment]
Sounds good! The project is here to stay, thank you for your contributions!
